src/qpx_mmc_css.cpp

```diff
@@ -0,0 +1,1909 @@
+        if (!((drive->media.dvdcss.regmask >> i) & 1)) {
+          //	printf("%d",i+1);
+          drive->rpc.region = i + 1;
+        }
+      //			printf("\n");
+      //		} else {
+      //			printf("Invalid region mask!\n");
+    }
+  }
+
+  return 0;
+}
+
+int read_disc_key(drive_info *drive, unsigned char DK[DVD_DISCKEY_SIZE]) {
+  int len = 2052; // 4 (header) + 2048 (key)
+  if (!(drive->rd_capabilities & DEVICE_DVD) ||
+      !(drive->capabilities & CAP_DVD_CSS))
+    return -1;
+
+  drive->cmd[0] = MMC_READ_DVD_STRUCTURE;
+  drive->cmd[7] = 0x02;
+  drive->cmd[8] = len >> 8;   //  -|
+  drive->cmd[9] = len & 0xFF; //  -- transfer length
+  drive->cmd[10] = drive->media.dvdcss.agid & 0xC0;
+  drive->cmd[11] = 0;
+  if ((drive->err = drive->cmd.transport(READ, drive->rd_buf, len))) {
+    if (!drive->silent)
+      sperror("READ_DISC_KEY", drive->err);
+    return drive->err;
+  }
+
+  memcpy(DK, drive->rd_buf + 4, 2048);
+  return 0;
+}
+
+int report_key(drive_info *drive, unsigned char key_class,
+               unsigned char key_format, int len, unsigned int lba) {
+  if (!(drive->rd_capabilities & DEVICE_DVD) ||
+      !(drive->capabilities & CAP_DVD_CSS))
+    return -1;
+
+  drive->cmd[0] = MMC_REPORT_KEY;
+
+  drive->cmd[2] = (lba >> 24) & 0xFF;
+  drive->cmd[3] = (lba >> 16) & 0xFF;
+  drive->cmd[4] = (lba >> 8) & 0xFF;
+  drive->cmd[5] = lba & 0xFF;
+
+  drive->cmd[7] = key_class;
+  drive->cmd[8] = len >> 8;
+  drive->cmd[9] = len & 0xFF;
+  drive->cmd[10] = (drive->media.dvdcss.agid & 0xC0) | (key_format & 0x3F);
+  drive->cmd[11] = 0;
+  if ((drive->err = drive->cmd.transport(READ, drive->rd_buf, len))) {
+    if (!drive->silent)
+      sperror("MMC REPORT KEY", drive->err);
+    return drive->err;
+  }
+  return 0;
+}
+
+int get_rpc_state(drive_info *drive) {
+  //	int len=8;
+  int i;
+  unsigned char regmask;
+  unsigned char t; //,vl,ul;
+  unsigned char sh;
+
+  if (!(drive->rd_capabilities & DEVICE_DVD) ||
+      !(drive->capabilities & CAP_DVD_CSS)) {
+    drive->rpc.phase = 0;
+    return -1;
+  }
+  /*
+      drive->cmd[0] = MMC_REPORT_KEY;
+      drive->cmd[7] = 0; // key class = 0 : DVD CSS/CPRM
+      drive->cmd[8] = len>>8;
+      drive->cmd[9] = len & 0xFF;
+      drive->cmd[10] = 0x08; // key format = 8 : RPC info
+      drive->cmd[11] = 0;
+  */
+  report_key(drive, 0, 0x08, 8);
+  if (drive->err) {
+    if (drive->err == 0x52400) {
+      drive->rpc.phase = 1;
+      return 0;
+    } else {
+      if (!drive->silent)
+        sperror("READ_RPC_STATE", drive->err);
+      drive->rpc.phase = 0;
+      return drive->err;
+    }
+  }
+  /*	printf("MMC_REPORT_KEY data: ");
+      for(i=0; i<len ;i++) printf(" %02X",drive->rd_buf[i] & 0xFF);
+      printf("\n");*/
+  if (ntoh16(drive->rd_buf) < 6)
+    return 1;
+
+  drive->rpc.ch_u = drive->rd_buf[4] & 0x07;
+  drive->rpc.ch_v = (drive->rd_buf[4] >> 3) & 0x07;
+  t = (drive->rd_buf[4] >> 6) & 0x03;
+  regmask = drive->rd_buf[5];
+  sh = drive->rd_buf[6];
```

Comment:
Codacy found an issue: [Variable 'sh' is assigned a value that is never used.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)